File: app/src/main/java/com/lucagiorgetti/surprix/ui/mainfragments/missingowners/MissingOwnersViewModel.java
package com.lucagiorgetti.surprix.ui.mainfragments.missingowners;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.lucagiorgetti.surprix.SurprixApplication;
import com.lucagiorgetti.surprix.listenerInterfaces.FirebaseListCallback;
import com.lucagiorgetti.surprix.model.User;
import com.lucagiorgetti.surprix.ui.BaseViewModel;
import com.lucagiorgetti.surprix.utility.dao.DoubleListDao;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class MissingOwnersViewModel extends BaseViewModel {
    private MutableLiveData<List<User>> missingOwners;

    public MissingOwnersViewModel(@NonNull Application application) {
        super(application);
        this.setLoading(false);
    }

    MutableLiveData<List<User>> getMissingOwners(String surpriseId) {
        if (missingOwners == null) {
            missingOwners = new MutableLiveData<>();
            loadOwners(surpriseId);
        }

        return missingOwners;
    }

    void loadOwners(String surpriseId) {
        new DoubleListDao(SurprixApplication.getInstance().getCurrentUser().getUsername()).getMissingOwners(surpriseId, new FirebaseListCallback<User>() {
            @Override
            public void onSuccess(List<User> users) {
                final ArrayList<User> owners = new ArrayList<>();
                final ArrayList<User> abroad_owners = new ArrayList<>();

                if (users != null){
                    for (User u : users) {
                        if (Objects.equals(u.getCountry(), SurprixApplication.getInstance().getCurrentUser().getCountry())) {
                            owners.add(u);
                        } else {
                            abroad_owners.add(u);
                        }
                    }
                }

                owners.addAll(abroad_owners);
                missingOwners.setValue(owners);
                setLoading(false);
            }

            @Override
            public void onStart() {
                setLoading(true);
            }

            @Override
            public void onFailure() {
                setLoading(false);
            }
        });

    }
}

File: app/src/main/java/com/lucagiorgetti/surprix/ui/mainfragments/missingowners/MissingOwnersViewModel.kt
package com.lucagiorgetti.surprix.ui.mainfragments.missingowners

import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.lucagiorgetti.surprix.SurprixApplication.Companion.getInstance
import com.lucagiorgetti.surprix.listenerInterfaces.FirebaseListCallback
import com.lucagiorgetti.surprix.model.User
import com.lucagiorgetti.surprix.ui.BaseViewModel
import com.lucagiorgetti.surprix.utility.dao.DoubleListDao

class MissingOwnersViewModel(application: Application) : BaseViewModel(application) {
    private var missingOwners: MutableLiveData<List<User>>? = null

    init {
        setLoading(false)
    }

    fun getMissingOwners(surpriseId: String?): MutableLiveData<List<User>> {
        if (missingOwners == null) {
            missingOwners = MutableLiveData()
            loadOwners(surpriseId)
        }
        return missingOwners!!
    }

    fun loadOwners(surpriseId: String?) {
        DoubleListDao(getInstance().currentUser?.username).getMissingOwners(surpriseId, object : FirebaseListCallback<User> {
            override fun onSuccess(users: MutableList<User>) {
                val owners = ArrayList<User>()
                val abroadOwners = ArrayList<User>()
                for (u in users) {
                    if (u.country == getInstance().currentUser?.country) {
                        owners.add(u)
                    } else {
                        abroadOwners.add(u)
                    }
                }
                owners.addAll(abroadOwners)
                missingOwners!!.value = owners
                setLoading(false)
            }

            override fun onStart() {
                setLoading(true)
            }

            override fun onFailure() {
                setLoading(false)
            }
        })
    }
}
